File: android/app/src/main/java/com/project/ahri/plugins/OverlayPlugin.java
package com.project.ahri.plugins;

import android.content.Context;
import android.content.Intent;
import android.net.Uri;
import android.os.Build;
import android.provider.Settings;
import androidx.core.content.ContextCompat;

import com.getcapacitor.JSObject;
import com.getcapacitor.Plugin;
import com.getcapacitor.PluginCall;
import com.getcapacitor.PluginMethod;
import com.getcapacitor.annotation.CapacitorPlugin;
import com.project.ahri.overlay.FloatingBubbleService;

@CapacitorPlugin(name = "Overlay")
public class OverlayPlugin extends Plugin {

    @PluginMethod
    public void checkPermission(PluginCall call) {
        JSObject ret = new JSObject();
        boolean granted = true;
        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.M) {
            granted = Settings.canDrawOverlays(getContext());
        }
        ret.put("granted", granted);
        call.resolve(ret);
    }

    @PluginMethod
    public void requestPermission(PluginCall call) {
        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.M) {
            if (!Settings.canDrawOverlays(getContext())) {
                Intent intent = new Intent(
                    Settings.ACTION_MANAGE_OVERLAY_PERMISSION,
                    Uri.parse("package:" + getContext().getPackageName())
                );
                if (getActivity() != null) {
                    getActivity().startActivity(intent);
                }
            }
        }
        call.resolve();
    }

    @PluginMethod
    public void showBubble(PluginCall call) {
        Context context = getContext();
        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.M && !Settings.canDrawOverlays(context)) {
            call.reject("Overlay permission not granted. Please enable 'Display over other apps' in Settings.");
            return;
        }
        Intent intent = new Intent(context, FloatingBubbleService.class);
        ContextCompat.startForegroundService(context, intent);
        JSObject ret = new JSObject();
        ret.put("success", true);
        call.resolve(ret);
    }

    @PluginMethod
    public void hideBubble(PluginCall call) {
        Context context = getContext();
        Intent intent = new Intent(context, FloatingBubbleService.class);
        context.stopService(intent);
        JSObject ret = new JSObject();
        ret.put("success", true);
        call.resolve(ret);
    }

    @PluginMethod
    public void isBubbleVisible(PluginCall call) {
        JSObject ret = new JSObject();
        ret.put("visible", FloatingBubbleService.isRunning);
        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.M) {
            ret.put("hasPermission", Settings.canDrawOverlays(getContext()));
        } else {
            ret.put("hasPermission", true);
        }
        call.resolve(ret);
    }
}

File: android/app/src/main/java/com/project/ahri/plugins/OverlayPlugin.kt
package com.project.ahri.plugins

import android.content.Intent
import android.net.Uri
import android.os.Build
import android.provider.Settings
import androidx.core.content.ContextCompat
import com.getcapacitor.JSObject
import com.getcapacitor.Plugin
import com.getcapacitor.PluginCall
import com.getcapacitor.PluginMethod
import com.getcapacitor.annotation.CapacitorPlugin
import com.project.ahri.overlay.FloatingBubbleService

@CapacitorPlugin(name = "Overlay")
class OverlayPlugin : Plugin() {
    
    @PluginMethod
    fun requestPermission(call: PluginCall) {
        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.M) {
            if (!Settings.canDrawOverlays(context)) {
                val intent = Intent(
                    Settings.ACTION_MANAGE_OVERLAY_PERMISSION,
                    Uri.parse("package:${context.packageName}")
                )
                activity.startActivityForResult(intent, 1001)
            }
        }
        call.resolve()
    }
    
    @PluginMethod
    fun showBubble(call: PluginCall) {
        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.M && !Settings.canDrawOverlays(context)) {
            call.reject("Overlay permission not granted")
            return
        }
        val intent = Intent(context, FloatingBubbleService::class.java)
        ContextCompat.startForegroundService(context, intent)
        call.resolve()
    }
    
    @PluginMethod
    fun hideBubble(call: PluginCall) {
        val intent = Intent(context, FloatingBubbleService::class.java)
        context.stopService(intent)
        call.resolve()
    }
    
    @PluginMethod
    fun isBubbleVisible(call: PluginCall) {
        val ret = JSObject()
        ret.put("visible", false)
        call.resolve(ret)
    }
}
